File: 1-100/23-merge-k-sorted-lists.cpp
struct Comp {
    bool operator() (ListNode* l1, ListNode* l2) {
        return l1->val > l2->val;
    }
};

ListNode* mergeKLists(vector<ListNode*>& lists) {
    if (lists.empty()) return nullptr;
    priority_queue<ListNode*, vector<ListNode*>, Comp> q;
    for (const auto &list : lists) {
        if (list != nullptr) {
            q.push(list);
        }
    }
    ListNode* dummyHead = new ListNode(-1), *curr = dummyHead;
    while (!q.empty()) {
        curr->next = q.top();
        q.pop();
        curr = curr->next;
        if (curr->next != nullptr) {
            q.push(curr->next);
        }
    }

    return dummyHead -> next;
}

File: freq_qa/23-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

ListNode.__lt__ = lambda a, b: a.val < b.val  # 让堆可以比较节点大小

class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        k = len(lists)
        if k == 0: return None
        h = [head for head in lists if head]
        heapify(h)
        dummy_head = ListNode()
        curr = dummy_head
        while h:
            node = heappop(h)
            if node.next:
                heappush(h, node.next)
            curr.next = node
            curr = curr.next
        return dummy_head.next
